src/utils/sortTypes.ts

```diff
@@ -0,0 +1,106 @@
+import { ScalarTypeComposer } from '../ScalarTypeComposer';
+import { EnumTypeComposer } from '../EnumTypeComposer';
+import { InterfaceTypeComposer } from '../InterfaceTypeComposer';
+import { InputTypeComposer } from '../InputTypeComposer';
+import { ObjectTypeComposer } from '../ObjectTypeComposer';
+import { UnionTypeComposer } from '../UnionTypeComposer';
+import { NamedTypeComposer } from './typeHelpers';
+import { SchemaFilterTypes } from './getFromSchema';
+import { isFunction } from './is';
+
+export type CompareTypeComposersResult = -1 | 0 | 1;
+
+export type CompareTypeComposersFn = (
+  tc1: NamedTypeComposer<any>,
+  tc2: NamedTypeComposer<any>,
+) => CompareTypeComposersResult;
+
+export type CompareTypeComposersOption =
+  | boolean
+  | 'ALPHABETIC'
+  | 'GROUP_BY_TYPE'
+  | CompareTypeComposersFn;
+
+const rootOrderDefault = [
+  'Query',
+  'Mutation',
+  'Subscription',
+];
+
+export function printSortAlpha(
+  tc1: NamedTypeComposer<any>,
+  tc2: NamedTypeComposer<any>,
+): CompareTypeComposersResult {
+  const comp = tc1.getTypeName().localeCompare(tc2.getTypeName());
+  return comp as CompareTypeComposersResult;
+}
+
+function sortGetPositionOfType(
+  tc: NamedTypeComposer<any>,
+  rootTypes: string[] = [],
+): number[] {
+  switch (true) {
+    case tc instanceof ScalarTypeComposer: return [2];
+    case tc instanceof EnumTypeComposer: return [3];
+    case tc instanceof UnionTypeComposer: return [4];
+    case tc instanceof InterfaceTypeComposer: return [5];
+    case tc instanceof ObjectTypeComposer:
+      const rootPos = rootTypes.indexOf(tc.getTypeName());
+      if (rootPos !== -1) {
+        return [1, rootPos];
+      } else {
+        return [6];
+      }
+    case tc instanceof InputTypeComposer: return [7];
+  }
+  throw new Error(`Unknown kind of type ${tc.getTypeName()}`);
+}
+
+function comparePositionLists(
+  p1: number[],
+  p2: number[],
+): CompareTypeComposersResult {
+  let common = Math.min(p1.length, p2.length);
+  for (let i = 0; i < common; i++) {
+    if (p1[i] < p2[i]) return -1;
+    if (p1[i] > p2[i]) return +1;
+  }
+  return 0;

```

Comment:
Yeah, that's a tricky one. I always like to make functions as generic as possible to enable reuse (or future modification), but in this case I guess it's not warranted 😅

As I said before, I'm not an expert on GraphQL, so I don't know if it's possible that one can have 10 or more root objects in the schema, in which case the sorting would break. That's the only reason I'd want to avoid a float.

But what do you think about this?

```ts
function comparePositionLists(
  p1: number[],
  p2: number[],
): CompareTypeComposersResult {
  if (p1[0] < p2[0]) return -1;
  if (p1[0] > p2[0]) return +1;
  if (p1[1] < p2[1]) return -1;
  if (p1[1] > p2[1]) return +1;
  return 0;
}
```